File: applications/rtkosem/rtkosem.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkosem_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkOSEMConeBeamReconstructionFilter.h"
#include "rtkPhaseGatingImageFilter.h"
#include "rtkIterationCommands.h"

#ifdef RTK_USE_CUDA
#  include "itkCudaImage.h"
#endif

#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtkosem, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;

#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, Dimension>;
#else
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;
#endif

  // Projections reader
  using ReaderType = rtk::ProjectionsReader<OutputImageType>;
  auto reader = ReaderType::New();
  rtk::SetProjectionsReaderFromGgo<ReaderType, args_info_rtkosem>(reader, args_info);

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::endl;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));

  // Create input: either an existing volume read from a file or a blank image
  itk::ImageSource<OutputImageType>::Pointer inputFilter;
  if (args_info.input_given)
  {
    // Read an existing image to initialize the volume
    auto inputReader = itk::ImageFileReader<OutputImageType>::New();
    inputReader->SetFileName(args_info.input_arg);
    inputFilter = inputReader;
  }
  else
  {
    // Create new empty volume
    using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
    auto constantImageSource = ConstantImageSourceType::New();
    rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkosem>(constantImageSource, args_info);
    constantImageSource->SetConstant(1.);
    inputFilter = constantImageSource;
  }

  // OSEM reconstruction filter
  auto osem = rtk::OSEMConeBeamReconstructionFilter<OutputImageType>::New();

  // Set the forward and back projection filters
  SetForwardProjectionFromGgo(args_info, osem.GetPointer());
  SetBackProjectionFromGgo(args_info, osem.GetPointer());
  osem->SetInput(inputFilter->GetOutput());
  osem->SetInput(1, reader->GetOutput());
  osem->SetGeometry(geometry);
  if (args_info.betaregularization_given)
    osem->SetBetaRegularization(args_info.betaregularization_arg);

  osem->SetNumberOfIterations(args_info.niterations_arg);
  osem->SetNumberOfProjectionsPerSubset(args_info.nprojpersubset_arg);
  osem->SetStoreNormalizationImages(!args_info.nostorenormalizationimages_flag);
  REPORT_ITERATIONS(osem, rtk::OSEMConeBeamReconstructionFilter<OutputImageType>, OutputImageType)

  // Write
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(osem->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkosem/rtkosem.py
#!/usr/bin/env python3
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Reconstructs a 3D volume from a sequence of projections with Ordered subset expectation maximization."
    )

    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--niterations", "-n", help="Number of iterations", type=int, default=5
    )
    parser.add_argument("--input", "-i", help="Input volume", type=str)
    parser.add_argument(
        "--nprojpersubset",
        help="Number of projections processed between each update of the reconstructed volume (several for OSEM, all for MLEM)",
        type=int,
        default=1,
    )
    parser.add_argument(
        "--betaregularization",
        help="Hyperparameter for the regularization",
        type=float,
        default=0.0,
    )
    parser.add_argument(
        "--nostorenormalizationimages",
        help="Do not store the normalization images during the reconstruction",
        action="store_true",
    )

    # RTK specific groups
    rtk.add_rtkinputprojections_group(parser)
    rtk.add_rtkprojectors_group(parser)
    rtk.add_rtk3Doutputimage_group(parser)

    return parser


def process(args_info: argparse.Namespace):
    # Define pixel type and dimension
    OutputPixelType = itk.F
    Dimension = 3
    InputImageType = itk.Image[OutputPixelType, Dimension]

    # Projections reader
    reader = rtk.ProjectionsReader[InputImageType].New()
    rtk.SetProjectionsReaderFromArgParse(reader, args_info)
    if args_info.verbose:
        print("Reading projections...")
    reader.Update()
    projections = reader.GetOutput()

    # Geometry
    if args_info.verbose:
        print(f"Reading geometry information from {args_info.geometry}...")
    geometry = rtk.read_geometry(args_info.geometry)

    # Create input: either an existing volume read from a file or a blank image
    if args_info.input:
        if args_info.verbose:
            print(f"Reading input volume from {args_info.input}...")
        inputFilter = itk.imread(args_info.input)
    else:
        # Create new empty volume
        constant = rtk.ConstantImageSource[InputImageType].New()
        rtk.SetConstantImageSourceFromArgParse(constant, args_info)
        constant.SetConstant(1.0)
        constant.Update()
        inputFilter = constant.GetOutput()

    # OSEM reconstruction filter
    if hasattr(itk, "CudaImage"):
        CudaImageType = itk.CudaImage[OutputPixelType, Dimension]
        osem = rtk.OSEMConeBeamReconstructionFilter[CudaImageType, CudaImageType].New()
        osem.SetInput(itk.cuda_image_from_image(inputFilter))
        osem.SetInput(1, itk.cuda_image_from_image(projections))
    else:
        osem = rtk.OSEMConeBeamReconstructionFilter[
            InputImageType, InputImageType
        ].New()
        osem.SetInput(inputFilter)
        osem.SetInput(1, projections)

    osem.SetGeometry(geometry)

    # Set the forward and back projection filters
    rtk.SetForwardProjectionFromArgParse(args_info, osem)
    rtk.SetBackProjectionFromArgParse(args_info, osem)

    if args_info.betaregularization is not None:
        osem.SetBetaRegularization(float(args_info.betaregularization))

    osem.SetNumberOfIterations(args_info.niterations)
    osem.SetNumberOfProjectionsPerSubset(args_info.nprojpersubset)
    osem.SetStoreNormalizationImages(not args_info.nostorenormalizationimages)

    # Report iterations
    progressCommand = rtk.PercentageProgressCommand(osem)
    osem.AddObserver(
        itk.ProgressEvent(), progressCommand.callback
    )  # Register the callback for progress
    osem.AddObserver(itk.EndEvent(), progressCommand.End)  # Register end notification

    # Write
    if args_info.verbose:
        print("Reconstructing and writing...")
    WriterType = itk.ImageFileWriter[InputImageType].New()
    WriterType.SetFileName(args_info.output)
    WriterType.SetInput(osem.GetOutput())
    WriterType.Update()


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
